### ISSUE TITLE
ttnn.transformer.scaled_dot_product_attention error scales with number of heads

### ISSUE THREAD

**Issue by philei-tt** [tenstorrent/tt-metal#20986]
### Describe the bug

The bigger the number of heads, the bigger difference compared to torch implementation. Here are some logs: 
```
input shape: (1, 8, 32, 32), pcc: 0.998494037987913, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(15.1250, dtype=torch.bfloat16)
input shape: (1, 8, 64, 32), pcc: 0.9977120095472902, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(25.7500, dtype=torch.bfloat16)
input shape: (1, 8, 128, 32), pcc: 0.9937824343052523, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(45.7500, dtype=torch.bfloat16)
input shape: (1, 8, 256, 32), pcc: 0.9841203640259895, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(90., dtype=torch.bfloat16)
input shape: (1, 8, 512, 32), pcc: 0.9625529834525473, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(204., dtype=torch.bfloat16)
input shape: (1, 8, 1024, 32), pcc: 0.8981233042666082, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(356., dtype=torch.bfloat16)
input shape: (1, 8, 2048, 32), pcc: 0.5273197165253803, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(536., dtype=torch.bfloat16)
input shape: (1, 8, 4096, 32), pcc: 0.2532025318287283, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(5248., dtype=torch.bfloat16)
input shape: (1, 8, 8192, 32), pcc: 0.10381338180604385, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(32256., dtype=torch.bfloat16)
input shape: (1, 8, 16384, 32), pcc: 0.06564090602237785, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(40960., dtype=torch.bfloat16)
```

I know that critical error appears for a huge number of heads, which we typically don't see in models, but if we fix that bug, I think it can improve accuracy a little bit for a smaller numbers of heads, too (64, 128, 256)

### Steps to reproduce the issue

Please use this script to reproduce logs from above: 

<details> 
<summary> Script (spoiler) </summary>

```
import torch
import ttnn
import ttnn

device_id = 0
dispatch_core_type = ttnn.device.DispatchCoreType.ETH
device = ttnn.open_device(
    device_id=device_id, l1_small_size=8192, dispatch_core_config=ttnn.device.DispatchCoreConfig(dispatch_core_type)
)
ttnn.enable_program_cache(device)

input_shapes = [
    (1, 8, 32, 32),
    (1, 8, 64, 32),
    (1, 8, 128, 32),
    (1, 8, 256, 32),
    (1, 8, 512, 32),
    (1, 8, 1024, 32),
    (1, 8, 2048, 32),
    (1, 8, 4096, 32),
    (1, 8, 8192, 32),
    (1, 8, 16384, 32),
    
    # (1, 8, 64, 32),
    # (1, 8, 64, 64),
    # (1, 8, 64, 128),
    # (1, 8, 64, 256),
    # (1, 8, 64, 512),
    # (1, 8, 64, 1024),
    # (1, 8, 64, 2048),
    # (1, 8, 64, 4096),     # Those are OOM (L1)
    # (1, 8, 64, 8192),     # Those are OOM (L1)
    # (1, 8, 64, 16384),    # Those are OOM (L1)

    # (1, 1, 64, 32),
    # (1, 4, 64, 32),
    # (1, 8, 64, 32),
    # (1, 16, 64, 32),
    # (1, 32, 64, 32),
    # (1, 64, 64, 32),
    # (1, 128, 64, 32),
    # (1, 256, 64, 32),
    # (1, 512, 64, 32),

]


compute_kernel_config = ttnn.WormholeComputeKernelConfig(
    math_fidelity=ttnn.MathFidelity.HiFi4,
    math_approx_mode=False,
    fp32_dest_acc_en=False,
    packer_l1_acc=False,
)


def test_attention():
    torch.manual_seed(0)
    for input_shape in input_shapes:
        query = torch.rand(input_shape, dtype=torch.bfloat16)
        key = torch.rand(input_shape, dtype=torch.bfloat16)
        value = torch.rand(input_shape, dtype=torch.bfloat16)

        tt_query = ttnn.from_torch(query, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)
        tt_key = ttnn.from_torch(key, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)
        tt_value = ttnn.from_torch(value, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)

        # scale = 10.0
        scale = None

        torch_out = torch.nn.functional.scaled_dot_product_attention(
            query, key, value, scale=scale, is_causal=False
        )
        ttnn_out = ttnn.transformer.scaled_dot_product_attention(
            tt_query,
            tt_key,
            tt_value,
            scale=scale,
            is_causal=False,
            compute_kernel_config=compute_kernel_config,
        )
        torch_ttnn_out = ttnn.to_torch(ttnn_out)

        pcc = ttnn.pearson_correlation_coefficient(
            torch_out,
            torch_ttnn_out,
        )
        print(
            f"input shape: {input_shape}, pcc: {pcc}, abs(sum(torch_out - torch_ttnn_out)): {torch.abs(torch.sum(torch_out - torch_ttnn_out))}"
        )


test_attention()

```
</details>

### Expected behavior

Accuracy should be 0.999+ for all possible shapes ideally

### Please complete the following environment information

- OS version: Ubuntu 24.04 LTS
- Python version: Python 3.10.17
- Commit: 5147622f25fc1cdb13f83cca1762b7f623cc2b00

**Comment by philei-tt** [comment]
@bbradelTT @yieldthought @cglagovichTT Please take a look

**Comment by cglagovichTT** [comment]
Can you refer to the other issue on how to set the program config to improve correctness?
https://github.com/tenstorrent/tt-metal/issues/20710

It looks like you're keeping heads constant at 8 and increasing sequence length. I have SDPA tests which test up to 128k with good PCC. Can you please set the program config with the hints I provided in the linked issue and try again?

**Comment by philei-tt** [comment]
@cglagovich  Ohh yeah you're right shape[3] is sequence length not head count. And yeah I'm sorry that I didn't notice you'r comment on last issue, you added it after I closed the issue :) 

Yeah tests are extremely useful. Seems like the issue was data generation. I was generating data in range [0:1], and it accuracy is bad for big sequences when there are no negative values in the input. 

Please close the issue if that's expected.

You can verify that with the next script: 

<details> 
<summary> script spoiler </summary>

```python
import torch
import ttnn
import ttnn

device_id = 0
dispatch_core_type = ttnn.device.DispatchCoreType.ETH
device = ttnn.open_device(
    device_id=device_id, l1_small_size=8192, dispatch_core_config=ttnn.device.DispatchCoreConfig(dispatch_core_type)
)
ttnn.enable_program_cache(device)

input_shapes = [
    (1, 8, 2048, 32),
]

def rand_in_range(shape, a, b, dtype=torch.bfloat16):
    rand_tensor = torch.rand(*shape, dtype=dtype) 
    return a + rand_tensor * (b - a)

def test_attention(a, b):
    torch.manual_seed(0)
    for input_shape in input_shapes:
        print(f"Data generated in range: [{a}:{b}]")
        query = rand_in_range(input_shape, a, b, dtype=torch.bfloat16)
        key = rand_in_range(input_shape, a, b, dtype=torch.bfloat16)
        value = rand_in_range(input_shape, a, b, dtype=torch.bfloat16)

        tt_query = ttnn.from_torch(query, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)
        tt_key = ttnn.from_torch(key, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)
        tt_value = ttnn.from_torch(value, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)
        
        # scale = 10.0
        scale = None

        torch_out = torch.nn.functional.scaled_dot_product_attention(
            query, key, value, scale=scale, is_causal=False, attn_mask=None
        )
        
        ttnn.device.DisablePersistentKernelCache()
        program_config = ttnn.SDPAProgramConfig(
            compute_with_storage_grid_size=device.compute_with_storage_grid_size(),
            q_chunk_size=256,
            k_chunk_size=256,
            exp_approx_mode=False,
        )

        compute_kernel_config = ttnn.WormholeComputeKernelConfig(
            math_fidelity=ttnn.MathFidelity.HiFi2,
            math_approx_mode=True,
            fp32_dest_acc_en=False,
            packer_l1_acc=False,
        )        
        ttnn_out = ttnn.transformer.scaled_dot_product_attention(
            tt_query,
            tt_key,
            tt_value,
            is_causal=False,
            attn_mask=None,
            compute_kernel_config=compute_kernel_config,
            program_config=program_config,
        )
        torch_ttnn_out = ttnn.to_torch(ttnn_out)

        pcc = ttnn.pearson_correlation_coefficient(
            torch_out,
            torch_ttnn_out,
        )
        print(
            f"input shape: {input_shape}, pcc: {pcc}, abs(sum(torch_out - torch_ttnn_out)): {torch.abs(torch.sum(torch_out - torch_ttnn_out))}"
        )


test_attention(0, 1)
test_attention(-1, 1)

```
</details>

Output: 
```
Data generated in range: [0:1]
input shape: (1, 8, 2048, 32), pcc: 0.8460759520545759, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(3152., dtype=torch.bfloat16)
Data generated in range: [-1:1]
input shape: (1, 8, 2048, 32), pcc: 0.9997609598979083, abs(sum(torch_out - torch_ttnn_out)): TorchTensor(18.7500, dtype=torch.bfloat16)
```

**Comment by cglagovichTT** [comment]
I'll close. Btw if you check the SDPA tests, you can see how we generate random inputs following the guide of [FA3](https://arxiv.org/pdf/2407.08608)
